### ISSUE TITLE
Validation key is not nested properly in 0.settings file

### ISSUE THREAD

**Issue by russellseymour** [chef-partners/azure-chef-extension#155]
Using the Chef Extension in an ARM template is generating an invalid `0.Settings` file where the validation key is not nested properly in `protectedSettings`.

The ChefExtension is being configured using the following template:

```json
{
      "name": "[concat(parameters('vmName'),'/ChefExtension')]",
      "type": "Microsoft.Compute/virtualMachines/extensions",
      "location": "[resourceGroup().location]",
      "apiVersion": "2015-06-15",
      "properties": {
        "publisher": "Chef.Bootstrap.WindowsAzure",
        "type": "ChefClient",
        "typeHandlerVersion": "1210.12",
        "settings": {
          "validation_key_format": "base64encoded",
          "bootstrap_options": {
            "chef_node_name": "[parameters('vmName')]",
            "chef_server_url": "[parameters('chefServerUrl')]",
            "validation_client_name": "[parameters('chefValidatorName')]",
            "node_ssl_verify_mode": "none"
          },
          "runlist": "[parameters('chefRunlist')]"
        },
        "protectedSettings": {
          "validation_key": "[parameters('chefValidatorKey')]"
        }
      }
    }
```

This generates the following `0.settings` file, which as can be seen has the key as the protectedSettings with no `validation_key` sub setting:

```json
{
  "runtimeSettings": [
    {
      "handlerSettings": {
        "protectedSettingsCertThumbprint": "D962D58667B0AB8B4CAAE669BE6139DBC40DC7D3",
        "protectedSettings": "MIITRAYJKoZIhvc...HFA9RK4pVow==",
        "publicSettings": {
          "validation_key_format": "base64encoded",
          "bootstrap_options": {
            "chef_node_name": "srh-node-1",
            "chef_server_url": "https://someserver123.northeurope.cloudapp.azure.com",
            "validation_client_name": "microsoft-validator",
            "node_ssl_verify_mode": "none"
          },
          "runlist": "recipe[awesome_customers_windows]"
        }
      }
    }
  ]
}
```

This makes the initial run fail as the `c:\chef\validation.pem` does not exist as it has not been extracted.

I have only tested this on Windows myself, I do not know if this is causing an issue on Linux but I suspect that it might.

/cc @jjasghar

**Comment by dheerajd-msys** [comment]
Hi @russellseymour, sub setting for `validation_key` in `protectedSettings` is not expected.
The waagent is populating the encrypted details in the  `protectedSettings`.
After Decrypting `protectedSettings` data, the decrypted data is in following key value pair format
```
"{\"validation_key\":\"<validation key>\"}"
``` 
If Base64 encoded validation key is passed in the template, then the `validation_key` value in the above key value pair contains the Base64 encoded validation key, which is base 64 decoded before populating `validation.pem`.
I checked by passing Base 64 encoded and plaintext validation key for Windows and Linux platofrm and in all the scenarios the `validation.pem` was getting created successfully and chef client run was successful.

**Comment by russellseymour** [comment]
@dheerajd-msys Thanks for this.

I will go back and do some testing as there is something not right.  The `validation.pem` file is not being created.  I know this is normally down to the encoding being incorrect.

I have seen this when running a workshop for Chef on Azure.  All of the delegates have Windows machines so I am generating the Base64 encoded string using the following:

```Powershell
PS > cd chef-repo
PS > $path = ".chef/acme-validator.pem"
PS > $Text = Get-Content -Path $path -Raw
PS > $Bytes = [System.Text.Encoding]::Unicode.GetBytes($Text)
PS > [Convert]::ToBase64String($Bytes)
```

This will create a string in the PS Console but it has carriage returns in it so people have to copy into a text editor and remove those.  Now it might be that everyone got this bit wrong and the extension was not able to decode the string because **all** of the labs failed.

So I will do some tests and try and work out what the problem is.

**Comment by NimishaS** [comment]
@russellseymour , please check if ruby's `Base64` module is helpful: http://ruby-doc.org/stdlib-2.1.3/libdoc/base64/rdoc/Base64.html#module-Base64-label-Example. We had used this only while testing.
One approach to avoid carriage return is redirecting the output of the command (that converts the input string into `Base64` encoded string) to a file, and then copy the string from the file.

**Comment by russellseymour** [comment]
@NimishaS I used the new 'Set-Clipboard` cmdlet in PowerShell 5 so that I had a string without carriage returns and it has been decoded properly in the `bootstrap.bat` file on the machine. 

I created a new Windows 2012 R2 machine with the Chef Extension and this is the error in the `C:\WindowsAzure\Logs\Plugins\Chef.Bootstrap.WindowsAzure.ChefClient\1210.12.105.1000\CommandExecution.txt` file:

```
[2016-07-25T16:49:48+00:00] WARN: chef-client configuration files creation failed (Expected process to exit with [0], but received '255'
---- Begin output of cmd.exe /C C:\Windows\TEMP/bootstrap.bat ----
STDOUT: Checking for existing directory "C:\chef"...
Existing directory found, skipping creation.
Writing validation key...
STDERR: ) was unexpected at this time.
---- End output of cmd.exe /C C:\Windows\TEMP/bootstrap.bat ----
Ran cmd.exe /C C:\Windows\TEMP/bootstrap.bat returned 255)
```

I checked the `bootstrap.bat` file and it contains the decoded validation key but there is a syntax error which means that the `validation.pem` file does not get written out properly thus `chef-client` fails to run properly.

However I downloaded the file to my local Windows 10 workstation and it ran without error so I suspect it is something that Windows 2012 R2 does not like.  Indeed if I run it locally on the server I get the same error:
 
![image](https://cloud.githubusercontent.com/assets/791658/17132650/9d5d1446-531a-11e6-9da1-8af4001ee531.png)

I will keep investigating but I suspect that Windows 2012 R2 does not like the way in which the content is being set for the file.

/cc @dheerajd-msys

**Comment by NimishaS** [comment]
@russellseymour , this seems like an issue with incorrect encoding of `validation.pem` file only. I verified again today with the given ruby script and it worked successfully for me:
```
require "base64"
text = File.read("D:/chef-repo/.chef/nimisha-validator.pem")
encoded_string = Base64.encode64(text)
```
I removed the newline characters from `encoded_string` and used it in the template. It worked fine. Can you please try this?

**Comment by russellseymour** [comment]
@NimishaS The problem is that not everyone is using Ruby to get the validation key in the correct format.  It needs to be possible to do it from PowerShell.

There are no carriage returns encoded or otherwise in the validation key.  Indeed the validation key is correct in the `bootstrap.bat` file meaning that it has been decoded correctly.  However I think there is a potential problem with the syntax in the batch file as I stated before.

If I copy the `bootstrap.bat` file to my **Windows 10** machine and run it then it works perfectly.  However when I run the same file manually on **Windows 2012 R2** the syntax error as pasted before is seen.

I do not think that there is an issue with using the following commands:

```
PS > cd chef-repo
PS > $path = ".chef/acme-validator.pem"
PS > $Text = Get-Content -Path $path -Raw
PS > $Bytes = [System.Text.Encoding]::Unicode.GetBytes($Text)
PS > [Convert]::ToBase64String($Bytes) | Set-Clipboard
```

If you run these yourself and then paste the contents of the clipboard into a text editor you will see that there are no carriage returns in the key.

Please can you try using PowerShell to create the base64 encoded string and then use the extension?  I appreciate it works with Ruby but we need it working with PowerShell as well - although I think we are looking at different issues here.

**Comment by NimishaS** [comment]
@russellseymour, I will try out with PS cmdlets.It's true that everyone will not be using ruby.

**Comment by russellseymour** [comment]
@NimishaS I meant to mention that the `Set-Clipboard` cmdlet is only available in PowerShell 5.  If you do this with PS 4 then you will have to redirect to a file and then copy out of that.

**Comment by dheerajd-msys** [comment]
Hi @russellseymour,
Thanks  for the debug.
As per the commands provided by you, i am able to reproduce the issue.
Working on fixing and testing the fix for various scenario's.

**Comment by dheerajd-msys** [comment]
@russellseymour,
I noticed that when we encode the key with Base64 in Powershell and decode it with ruby, we get the key with each character space separated. the validation key obtained [here](https://github.com/chef-partners/azure-chef-extension/blob/master/lib/chef/azure/commands/enable.rb#L285) is in ascii format , so after deleting the space, it works properly and validation.pem gets generated successfully.
This explains why it is working for ruby encoded validation key and fails for powershell encoded validation key. 
PR:- https://github.com/chef-partners/azure-chef-extension/pull/157

**Comment by russellseymour** [comment]
@dheerajd-msys Thats great - well spotted :-)

Can you let me know when this is likely to get released please?

**Comment by NimishaS** [comment]
@russellseymour , we are testing this change on various platforms with both `plaintext` and `base64encoded` validation key.  We can publish the extension in next week after discussion with @btm .

**Comment by dheerajd-msys** [comment]
@russellseymour, we have released the extension version `1210.12.105.1001` externally with the validation.pem fix

**Comment by russellseymour** [comment]
@dheerajd-msys @NimishaS I have now tested this with the code snippets above on Windows and it works.  Thanks for sorting this out.
